### DISCUSSION TITLE
OpenSCAD Language support

### DISCUSSION THREAD

**Discussion by csicar** [zalo/CascadeStudio#42] [Ideas]
Hi, is there interest in having a OpenSCAD as an additional input language? If so, I might be interested in implementing it.
That could be useful for reusing existing libraries. 

The rigid language of OpenSCAD could also allow for awesome code introspection. For example highlighting the line in the code, that generated the geometry.

**Comment by zalo** [comment]
Yeah; I'd love to see OpenSCAD as an optional language in CascadeStudio!

[In theory, this project (from the folks at OpenJSCAD) gets one most of the way to CascadeStudio compatibility...](https://github.com/jscad/openscad-openjscad-translator)   There may have to be a UI or detection step of some sort to switch between the input languages.

Unfortunately, code introspection is going to be a little more difficult, since it's based on Microsoft's powerful Typescript/Javascript analyzers.

However, if the one-time automatic conversion-step can be handled when an OpenSCAD script is dragged onto CascadeStudio, then the _converted code_ would be able to benefit from the introspection that CascadeStudio's Java/Typescript enjoys!

**Reply by csicar** [reply]
Yes, code and 3d-model introspection will need to implemented on our own, I don't think it should be the first priority, but sth to keep in mind.

I think I would prefer *interpreting* instead of *compiling* OpenSCAD code. That way it would leave room for implementing advanced introspection. E.g. When the user clicks on a piece of the model, the editor could jump to the place where it is defined in code, which would not be possible in plain JS.

One nice extra we would get from interpreting: We could expose a `EvalOpenScad` function to the JS environment. It would take an OpenSCAD source-string as input and return the corresponding module.
